### PR TITLE
Carla camera streaming integration

### PR DESCRIPTION
**Pull request by DKlimenko**
Carla camera streaming integration.

### CODE REVIEW

**Comment by arminn** [comment]
@svlad-90 @dterletskiy plz review

**Comment by svlad-90** [comment]
Hi @DKlimenko, please:

- add "Signed-off-by: <dmytro_klimenko@epam.com>" to your commit messages
- add a proper commit message to each commit which describes what was changed and the motivation for the change

File: evs.mk [hunk 1]
```diff
@@ -10,9 +6,11 @@ PRODUCT_PACKAGES += \
     android.hardware.automotive.evs@1.1.xt \
     android.automotive.evs.manager@1.1 \
     evs_app_xt \
-    android.hardware.automotive.evs@1.1-xt
+    android.hardware.automotive.evs@1.1-xt \
```
Comment (svlad-90) [inline evs.mk:6]: Hi @DKlimenko, do you really need this change? It is not functional. If really needed, please, move it to a separate commit.

File: sampleDriver/Android.bp [hunk 2]
```diff
@@ -61,11 +61,7 @@ cc_binary {
         "libui",
         "libutils",
         "libcutils",
-        "libyuv",
-        "libuws",
```
Comment (svlad-90) [inline sampleDriver/Android.bp]: Hi @DKlimenko, you've added some of these libs in a separate commit of this same PR - https://github.com/xen-troops/android_vendor_xen_evs/pull/9/commits/8f8a0105e307e39185aac7f7036a2ad4d2829716#diff-d84cecf0528f7adf07b53b01cb25d3a78aeca1e7ab0606d73204252687bf2363R65. Here you remove them.

Please, remove the "experiments" and leave only atomic valuable changes, without "fixing the bugs".

**Review by svlad-90 — CHANGES_REQUESTED** [review]

File: config_override.json [hunk 3]
```diff
@@ -12,63 +11,63 @@
...
-      "vfov" : 80,
-      "hflip" : false,
+      "hfov" : 80,
+      "vfov" : 15,
+      "hflip" : true,
       "vflip" : false
     },
-    {
-      "cameraId" : "/dev/video2",
+    {
```
Comment (svlad-90) [inline config_override.json]: Hi @DKlimenko, please, remove trailing spaces.

File: sampleDriver/EvsStreamingCamera.cpp [hunk 4]
```diff
@@ -73,9 +74,53 @@ EvsStreamingCamera::EvsStreamingCamera(const char *deviceName, unique_ptr<Config
               GRALLOC_USAGE_SW_WRITE_OFTEN;
 
     char propValue[PROPERTY_VALUE_MAX] = {};
-    property_get("persist.vendor.evs.uri", propValue, "ws://192.168.122.1:8888");
+    property_get("persist.vendor.evs.uri", propValue, "ws://192.168.122.1");
     mUri = propValue;
-    LOG(DEBUG) << "EvsStreamingCamera instantiated " << mUri;
+
+    if (!std::string(deviceName).compare("CarlaSimFront")) {
+        mUri += ":8887";
```
Comment (svlad-90) [inline sampleDriver/EvsStreamingCamera.cpp:81]: Hi @DKlimenko, should these ports be hard-coded? Or, maybe, it is worth making them separate properties?

File: sampleDriver/IEvsDriverCamera.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
+#ifndef ANDROID_HARDWARE_AUTOMOTIVE_EVS_V1_1_IEVSDRIVERCAMERA_H
```
Comment (svlad-90) [inline sampleDriver/IEvsDriverCamera.h:1]: Hi @DKlimenko, I do not see anyone using this interface. Is it really needed as part of this change?

**Comment by DKlimenko** [comment]
> Hi @DKlimenko, please:
> 
> * add "Signed-off-by: [dmytro_klimenko@epam.com](mailto:dmytro_klimenko@epam.com)" to your commit messages
> * add a proper commit message to each commit which describes what was changed and the motivation for the change

Hi. Squashed commits.
Was thinking PR supports squashing.
Signed-off added

File: sampleDriver/EvsStreamingCamera.cpp (hunk 4 above)
Comment (DKlimenko) [reply sampleDriver/EvsStreamingCamera.cpp:81]: No sure regarding properties as for now.

**Comment by svlad-90** [comment]
Hi @DKlimenko, I see signed-off-by only in https://github.com/xen-troops/android_device_xenvm/pull/92/commits/729273888cf2bef4791340d9aa4d5645a08415a3, but not in https://github.com/xen-troops/android_vendor_xen_evs/pull/9/commits/8e245b23a1e58d348424259969012bb92beb7b5f.

Also in both PRs, there are no commit messages, which would describe the changes.

**Comment by DKlimenko** [comment]
> Hi @DKlimenko, I see signed-off-by only in [xen-troops/android_device_xenvm@7292738](https://github.com/xen-troops/android_device_xenvm/commit/729273888cf2bef4791340d9aa4d5645a08415a3), but not in [8e245b2](https://github.com/xen-troops/android_vendor_xen_evs/commit/8e245b23a1e58d348424259969012bb92beb7b5f).
> 
> Also in both PRs, there are no commit messages, which would describe the changes.

Done

File: evs.mk [hunk 6]
```diff
@@ -1,4 +1,9 @@
-PRODUCT_PACKAGES += android.hardware.automotive.evs@1.1-xt
+#PRODUCT_PROPERTY_OVERRIDES += persist.automotive.evs.mode=1
```
Comment (arminn) [inline evs.mk]: Do we need that ?

File: sampleDriver/EvsEnumerator.cpp [hunk 7]
```diff
@@ -203,6 +205,16 @@ void EvsEnumerator::enumerateCameras() {
...
 
+    const std::string kCarlaSimReverse = "CarlaSimReverse";
+    const std::string kCarlaSimFront   = "CarlaSimFront";
+    const std::string kCarlaSimLeft    = "CarlaSimLeft";
+    const std::string kCarlaSimRight   = "CarlaSimRight";
+    sCameraList.emplace(kCarlaSimReverse, kCarlaSimReverse.c_str());
+    sCameraList.emplace(kCarlaSimFront, kCarlaSimFront.c_str());
+    sCameraList.emplace(kCarlaSimLeft,  kCarlaSimLeft.c_str());
+    sCameraList.emplace(kCarlaSimRight, kCarlaSimRight.c_str());
+    captureCount +=4;
```
Comment (arminn) [inline sampleDriver/EvsEnumerator.cpp:217]: I prefer a comment here to describe that we have some kind static configuration/expectations here.

File: sampleDriver/EvsEnumerator.cpp [hunk 8]
```diff
@@ -282,33 +294,52 @@ Return<sp<IEvsCamera_1_0>> EvsEnumerator::openCamera(const hidl_string& cameraId
     }
 
     // Is this a recognized camera id?
-    CameraRecord *pRecord = findCameraById(cameraId);
+    CameraRecord* pRecord = findCameraById(cameraId);
```
Comment (arminn) [inline sampleDriver/EvsEnumerator.cpp]: this change not needed

File: sampleDriver/EvsEnumerator.cpp [hunk 9]
```diff
@@ -282,33 +294,52 @@ Return<sp<IEvsCamera_1_0>> EvsEnumerator::openCamera(const hidl_string& cameraId
...
     if (pRecord == nullptr) {
         LOG(ERROR) << cameraId << " does not exist!";
         return nullptr;
     }
+ 
+    if (string::npos != std::string(cameraId.c_str()).find(kStreamingCamera)) {
+        LOG(WARNING) << "Create streaming camera";
```
Comment (arminn) [inline sampleDriver/EvsEnumerator.cpp]: Log info ?

File: sampleDriver/EvsEnumerator.cpp [hunk 10]
```diff
@@ -481,33 +513,55 @@ Return<sp<IEvsCamera_1_1>> EvsEnumerator::openCamera_1_1(const hidl_string& came
...
     }
 
-    // Has this camera already been instantiated by another caller?
-    sp<EvsV4lCamera> pActiveCamera = pRecord->activeInstance.promote();
-    if (pActiveCamera != nullptr) {
-        LOG(WARNING) << "Killing previous camera because of new caller";
-        closeCamera(pActiveCamera);
-    }
+    if (string::npos != std::string(cameraId.c_str()).find(kStreamingCamera)) {
+        LOG(WARNING) << "Create streaming camera";
```
Comment (arminn) [inline sampleDriver/EvsEnumerator.cpp]: Log info ?

File: sampleDriver/EvsEnumerator.h [hunk 11]
```diff
@@ -59,16 +60,13 @@ class EvsEnumerator : public IEvsEnumerator {
 
     // Methods from ::android::hardware::automotive::evs::V1_1::IEvsEnumerator follow.
     Return<void>                getCameraList_1_1(getCameraList_1_1_cb _hidl_cb) override;
-    Return<sp<IEvsCamera_1_1>>  openCamera_1_1(const hidl_string& cameraId,
-                                               const Stream& streamCfg) override;
+    Return<sp<IEvsCamera_1_1>>  openCamera_1_1(const hidl_string& cameraId, const Stream& streamCfg) override;
```
Comment (arminn) [inline sampleDriver/EvsEnumerator.h]: this change not needed

File: sampleDriver/EvsEnumerator.h [hunk 12]
```diff
@@ -59,16 +60,13 @@ class EvsEnumerator : public IEvsEnumerator {
...
     Return<bool>                isHardware() override { return true; }
     Return<void>                getDisplayIdList(getDisplayIdList_cb _list_cb) override;
     Return<sp<IEvsDisplay_1_1>> openDisplay_1_1(uint8_t port) override;
     Return<void> getUltrasonicsArrayList(getUltrasonicsArrayList_cb _hidl_cb) override;
-    Return<sp<IEvsUltrasonicsArray>> openUltrasonicsArray(
-            const hidl_string& ultrasonicsArrayId) override;
-    Return<void> closeUltrasonicsArray(
-            const ::android::sp<IEvsUltrasonicsArray>& evsUltrasonicsArray) override;
+    Return<sp<IEvsUltrasonicsArray>> openUltrasonicsArray(const hidl_string& ultrasonicsArrayId) override;
+    Return<void> closeUltrasonicsArray(const ::android::sp<IEvsUltrasonicsArray>& evsUltrasonicsArray) override;
```
Comment (arminn) [inline sampleDriver/EvsEnumerator.h]: this change not needed

File: sampleDriver/EvsEnumerator.h [hunk 13]
```diff
@@ -81,19 +79,18 @@ class EvsEnumerator : public IEvsEnumerator {
...
-        wp<EvsV4lCamera>    activeInstance;
+        CameraDesc                desc;
+        wp<EvsV4lCamera>          activeInstance;
+        wp<EvsStreamingCamera>    activeStreamingInstance;
 
         CameraRecord(const char *cameraId) : desc() { desc.v1.cameraId = cameraId; }
     };
 
     bool checkPermission();
-
```
Comment (arminn) [inline sampleDriver/EvsEnumerator.h]: this change not needed

File: sampleDriver/EvsEnumerator.h [hunk 14]
```diff
@@ -81,19 +79,18 @@ class EvsEnumerator : public IEvsEnumerator {
...
     static bool qualifyCaptureDevice(const char* deviceName);
     static CameraRecord* findCameraById(const std::string& cameraId);
     static void enumerateCameras();
     static void enumerateDisplays();
-
```
Comment (arminn) [inline sampleDriver/EvsEnumerator.h]: this change not needed

File: sampleDriver/EvsEnumerator.h [hunk 15]
```diff
@@ -102,21 +99,15 @@ class EvsEnumerator : public IEvsEnumerator {
...
-    static uint64_t                           sInternalDisplayId;
+    static std::unordered_map<std::string, CameraRecord> sCameraList;
+    static wp<EvsGlDisplay>                              sActiveDisplay; // Weak pointer.
+                                                                       // Object destructs if client dies.
+    static std::mutex                                    sLock;          // Mutex on shared camera device list.
+    static std::condition_variable                       sCameraSignal;  // Signal on camera device addition.
+    static std::unique_ptr<ConfigManager>                sConfigManager; // ConfigManager
+    static sp<IAutomotiveDisplayProxyService>            sDisplayProxy;
+    static std::unordered_map<uint8_t, uint64_t>         sDisplayPortList;
+    static uint64_t                                      sInternalDisplayId;
```
Comment (arminn) [inline sampleDriver/EvsEnumerator.h]: this changes not needed.

**Review by arminn — CHANGES_REQUESTED** [review]
I've created a branch android-12-carla-master plz change branch in PR. Also you need meaningful description for the patch. Pay attention: plz remove unneeded style changes, which no add/change functionality inside pull.

File: sampleDriver/EvsEnumerator.cpp (hunk 8 above)
Comment (DKlimenko) [reply sampleDriver/EvsEnumerator.cpp]: google doesn't have coding style, pointers and references are without rules

Comment (arminn) [reply sampleDriver/EvsEnumerator.cpp]: https://google.github.io/styleguide/cppguide.html  In any means - we stick to the original style of the file and don't change it. And not adding non functional style changes.

File: sampleDriver/EvsEnumerator.h (hunk 15 above)
Comment (DKlimenko) [reply sampleDriver/EvsEnumerator.h]: can't agree. it is our implementation and 
I would like to make it more readable.
lets discuss coding style

Comment (DKlimenko) [reply sampleDriver/EvsEnumerator.h]: done

File: sampleDriver/EvsEnumerator.h (hunk 14 above)
Comment (DKlimenko) [reply sampleDriver/EvsEnumerator.h]: done

File: sampleDriver/EvsEnumerator.h (hunk 13 above)
Comment (DKlimenko) [reply sampleDriver/EvsEnumerator.h]: done

File: sampleDriver/EvsEnumerator.h (hunk 12 above)
Comment (DKlimenko) [reply sampleDriver/EvsEnumerator.h]: done

File: sampleDriver/EvsEnumerator.h (hunk 11 above)
Comment (DKlimenko) [reply sampleDriver/EvsEnumerator.h]: done

**Review by arminn — APPROVED** [review]
